crates/wasi-provider/src/lib.rs

```diff
@@ -83,6 +90,127 @@ impl<S: Store + Send + Sync> WasiProvider<S> {
             kubeconfig,
         })
     }
+
+    fn volume_path_map(
+        container: &KubeContainer,
+        volumes: &HashMap<String, Ref>,
+    ) -> anyhow::Result<HashMap<PathBuf, Option<PathBuf>>> {
+        if let Some(volume_mounts) = container.volume_mounts.as_ref() {
+            volume_mounts
+                .iter()
+                .map(|vm| -> anyhow::Result<(PathBuf, Option<PathBuf>)> {
+                    // Check the volume exists first
+                    let vol = volumes.get(&vm.name).ok_or_else(|| {
+                        anyhow::anyhow!(
+                            "no volume with the name of {} found for container {}",
+                            vm.name,
+                            container.name
+                        )
+                    })?;
+                    let mut guest_path = PathBuf::from(&vm.mount_path);
+                    if let Some(sub_path) = &vm.sub_path {
+                        guest_path.push(sub_path);
+                    }
+                    // We can safely assume that this should be valid UTF-8 because it would have
+                    // been validated by the k8s API
+                    Ok((vol.deref().clone(), Some(guest_path)))
+                })
+                .collect::<anyhow::Result<HashMap<PathBuf, Option<PathBuf>>>>()
+        } else {
+            Ok(HashMap::default())
+        }
+    }
+
+    async fn run_one_container(
+        &self,
+        container: &KubeContainer,
+        pod: &Pod,
+        client: &kube::client::Client,
+        modules: &mut HashMap<String, Vec<u8>>,
+        volumes: &HashMap<String, Ref>,
+        container_handles: &mut HashMap<
```

Comment:
I think it would probably be better to just return a container handle from this function rather that passing a mutable handle. From a code organization standpoint I think it is clearer to manage the full map of handles in the main handler function (`add`)

Reply:
That sounds desirable.  Need to figure out how it will play with removing completed init containers from the map (which I found I had to do and am probably doing in the wrong place at the moment).